### PR TITLE
fix: Text splitting in the BasicTokenizer

### LINKED ISSUE #22166: Mismatch between CLIP fast and non-fast tokenizers
### System Info

- `transformers` version: 4.26.1
- Platform: Windows-10-10.0.22621-SP0
- Python version: 3.8.1
- Huggingface_hub version: 0.11.0
- PyTorch version (GPU?): 1.12.1+cu116 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@ArthurZucker @younesbelkada @sgugger

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Here is a minimal working example to show the mismatch:

```python
from transformers import AutoTokenizer

tokenizer_fast = AutoTokenizer.from_pretrained('openai/clip-vit-base-patch16', use_fast=True)
tokenizer = AutoTokenizer.from_pretrained('openai/clip-vit-base-patch16', use_fast=False)

text = "You should've done this"

print(tokenizer(text))
print(tokenizer_fast(text))
print(tokenizer(text) == tokenizer_fast(text))

# Outputs:
# {'input_ids': [49406, 592, 1535, 262, 563, 1700, 589, 49407], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1]}
# {'input_ids': [49406, 592, 1535, 1200, 1700, 589, 49407], 'attention_mask': [1, 1, 1, 1, 1, 1, 1]}
# False
```

It appears to stem from the `'ve` token.

### Expected behavior

The non-fast tokenization should match the fast tokenization (https://github.com/huggingface/tokenizers)

### PR DESCRIPTION
**Pull request by connor-henderson**
# What does this PR do?

Note: The tests are failing because of repo-consistency, but I intentionally haven't made the change across the repo until we confirm what change we want to make. Will remove [Don't merge] tag once done.

This fixes an issue related to the BasicTokenizer.

Initially looked to fix #22166 by updating the _run_split_on_punc method in the BasicTokenizer to split apostrophes without starting a new word. In the issue, it was noted that apostrophes weren't being split properly as `should've` was being converted to `should`, `'`, and `ve` instead of `should` and `'ve`.

However, when adding testing it became apparent there are other cases where the BasicTokenizer is failing too, such as capturing '!!' as separate tokens with id 256 as opposed to one 748 token, which is in the [vocab](https://huggingface.co/openai/clip-vit-base-patch32/resolve/main/vocab.json).

To address these I modified `_run_split_on_punc` in the BasicTokenizer to also split on passed patterns and renamed it `_split_on_punc_or_pattern` and added tests for them. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

File: src/transformers/models/bert/tokenization_bert.py
Comment (connor-henderson) [inline src/transformers/models/bert/tokenization_bert.py:456]: We could add the following line in the class init:
`self.pattern = pattern if pattern else r"([\-\[\]\(\)\{\}\.,!?;:''^$`‘’“”\"/\\]+)"`

We then wouldn't need the `if self.pattern` conditional or anything below this line in this fn, but there could be something about splitting punctuation I'm missing here.

What do you think about this idea?

**Comment by connor-henderson** [comment]
> I think this was still a fix we were interesting in having for users who don't have `ftfy` installed.

Oh ok reopening in that case

**Review by ArthurZucker — COMMENTED** [review]
Good work there! My main comment is to add the pattern in the init, and try to be more informative. So something like `self.tokenization_splitting_pattern` or something that is easy to understand since it is going to be in the `tokenizer_config.json`!

**Comment by connor-henderson** [comment]
Thanks for your comments Arthur! Looking back at it to make your changes I realized two things:
- we actually don't need to add the pattern splitting to the BasicTokenizer, we just need to *not* split on punctuation (just need to get out of the way of the byte pair encoding happening later)
- the already written [test_check_encoding_slow_fast](url) test is a bit more comprehensive, so I used that and it passes locally now without ftfy, I'll call out the additional edits I had to make below for this

Will wait for your review again before replicating it in the other BasicTokenizers. 

Also, would you like me to add additional testing? I.e., should tokenizers that use BasicTokenizer each have a test that checks slow and fast match like the comprehensive check like the one mentioned above that's in CLIP, or should they have a simple truncated version of it

File: src/transformers/models/clip/tokenization_clip.py
Comment (connor-henderson) [inline src/transformers/models/clip/tokenization_clip.py:133]: CLIP's vocab includes control chars, for example for [these two characters in the test](https://github.com/huggingface/transformers/blob/ae5fc2db87f2fb715c137c4823647f5b24c7f881/tests/models/clip/test_tokenization_clip.py#L104-L105)

File: src/transformers/models/bert/tokenization_bert.py
Comment (connor-henderson) [inline src/transformers/models/bert/tokenization_bert.py:435]: As the comment mentions, done for [this part](https://github.com/huggingface/transformers/blob/ae5fc2db87f2fb715c137c4823647f5b24c7f881/tests/models/clip/test_tokenization_clip.py#L90-L94) of the test

**Comment by ArthurZucker** [comment]
Okay! I'll review again, can you make sure `make quality` and `make repo-consistency` both pass?

**Comment by connor-henderson** [comment]
Hey @ArthurZucker just checking in, anything else wanted here?

**Comment by ArthurZucker** [comment]
Nope thanks for the ping, it is just that it is a lot of changes on a lot of models (a lot of old models too 😉 ). Getting to it!

File: tests/models/bert/test_tokenization_bert.py
Comment (ArthurZucker) [inline tests/models/bert/test_tokenization_bert.py:185]: nice thanks for adding tests!

**Review by ArthurZucker — APPROVED** [review]
This looks very clean to me. Thanks for working on that. Pinging @Narsil for a second pair of eyes. 
I think basically nothing changes for default behaviour and this fixes clip

**Review by Narsil — COMMENTED** [review]
I feel very uneasy about the NFC addition line.

It might be correct, but it's definitely a very sizeable change in behavior (which is not very visible in English).

File: tests/models/bert/test_tokenization_bert.py
Comment (connor-henderson) [reply tests/models/bert/test_tokenization_bert.py:188]: Sure, I'll add a test for when split on punc is False. This doesn't clash with the comment as I mentioned in the second iteration of this pr that I realized
>we actually don't need to add the pattern splitting to the BasicTokenizer, we just need to not split on punctuation (just need to get out of the way of the byte pair encoding happening later)

Essentially getting out of the way of [this](https://github.com/huggingface/transformers/blob/abbc96a2148da0c91fa078bd021984b2cc10ef85/src/transformers/models/clip/tokenization_clip.py#L474-L478) part

File: src/transformers/models/tapas/tokenization_tapas.py
Comment (connor-henderson) [reply src/transformers/models/tapas/tokenization_tapas.py:2105]: Happy to add a new path, maybe `do_normalize` that's default false, or even ditch this part. This line was added to satisfy one more edge-case like part of test [here](https://github.com/huggingface/transformers/blob/abbc96a2148da0c91fa078bd021984b2cc10ef85/tests/models/clip/test_tokenization_clip.py#L92-L96), so we'd still get the thrust of the pr without it. Just lmk what's preferred.

I believe it at least wouldn't be an issue for CLIP since ftfy also does this by default on the text.

Comment (Narsil) [reply src/transformers/models/tapas/tokenization_tapas.py:2105]: I will let core maintainers take a stance on which direction we should go.

I think adding at least a test to show normalization happens is necessary. The link you provide https://github.com/huggingface/transformers/blob/abbc96a2148da0c91fa078bd021984b2cc10ef85/tests/models/clip/test_tokenization_clip.py#L92-L96 does not showcase that. It checks that both treat this utf-8 accentuation in the same way, not that normalization happens (which would required encoding the utf-8 accentuation and the normalized form and check they have the same id, and ideally explicitly mentionning the ID within the test).

File: src/transformers/models/clip/tokenization_clip.py
Comment (connor-henderson) [reply src/transformers/models/clip/tokenization_clip.py]: ^update: it doesn't include them, those two in the test were just for matching the fast tokenizer which marks them as UNK

**Comment by Narsil** [comment]
Looking at the output for `ar` it seems  NEW + normalize is the best match isn't it ?

I think this proves that `NFC` is indeed a good addition which was previously missing ! 

Thanks a lot for this testing script, this adds a lot of value for future potential changes !

**Comment by connor-henderson** [comment]
>I think this proves that NFC is indeed a good addition which was previously missing !

Thanks and sounds good, I'll put it back in. I had removed it only because I couldn't test all languages due to the time it would take so I wasn't certain if there could be other issues with it, and the improvements were somewhat modest.

**Comment by connor-henderson** [comment]
Hey @Narsil just checking in, anything else wanted here?

**Comment by Narsil** [comment]
No we just need a core maintainer's approval.

Sorry I forgot about this PR.

@sgugger for final review.

**Review by sgugger — APPROVED** [review]
This looks good to me but I'm not following if you approve or not the changes in the end @Narsil , as there are some pretty strong unresolved comments :-)

**Comment by connor-henderson** [comment]
Noticed the linked issue was marked stale, this PR probably will be soon too. Any other action wanted here? 

I think as the script shows this will significantly improve how well the Basic Tokenizer matches up with the fast one for CLIP, the lingering question was just around whether the NFC normalizing change was approved or whether that part should be removed. @Narsil @sgugger

**Comment by sgugger** [comment]
Just waited for a clarification on whether @Narsil was fine with merging this or not. :-)

**Comment by sgugger** [comment]
Since @Narsil is not commenting, I'm guessing we can merge :man_shrugging:

**Pull request merged**
